lib/Dialect/AIEVec/IR/AIEVecOps.cpp

```diff
@@ -1620,8 +1627,105 @@ ParseResult ShiftOp::parse(OpAsmParser &parser, OperationState &result) {
+  case ShuffleMode::T8_16X4: // 36
+  case ShuffleMode::T8_4X16: // 37
+  case ShuffleMode::T8_8X4:  // 46
+  case ShuffleMode::T8_4X8:  // 47
+    requireRhs = false;
+    LLVM_FALLTHROUGH;
+  case ShuffleMode::T8_64X2_LO: //  0
+  case ShuffleMode::T8_64X2_HI: //  1
+  case ShuffleMode::T8_2X64_LO: // 20
+  case ShuffleMode::T8_2X64_HI: // 21
+    modeBitWidth = 8u;
+    break;
+  case ShuffleMode::T16_8X4:      // 28
+  case ShuffleMode::T16_4X8:      // 29
+  case ShuffleMode::T16_1X2_flip: // 38
+  case ShuffleMode::T16_4X4:      // 39
+  case ShuffleMode::T16_4X2:      // 40
+  case ShuffleMode::T16_2X4:      // 41
+  case ShuffleMode::T16_8X2:      // 42
+  case ShuffleMode::T16_2X8:      // 43
+  case ShuffleMode::T16_16X2:     // 44
+  case ShuffleMode::T16_2X16:     // 45
+    requireRhs = false;
+    LLVM_FALLTHROUGH;
+  case ShuffleMode::T16_32X2_LO: //  2
+  case ShuffleMode::T16_32X2_HI: //  3
+  case ShuffleMode::T16_2X32_LO: // 18
+  case ShuffleMode::T16_2X32_HI: // 19
+  case ShuffleMode::T16_16X4_LO: // 24
+  case ShuffleMode::T16_16X4_HI: // 25
+  case ShuffleMode::T16_4X16_LO: // 26
+  case ShuffleMode::T16_4X16_HI: // 27
+    modeBitWidth = 16u;
+    break;
+  case ShuffleMode::T32_4X4: // 34
+    requireRhs = false;
+    LLVM_FALLTHROUGH;
+  case ShuffleMode::T32_16X2_LO: //  4
+  case ShuffleMode::T32_16X2_HI: //  5
+  case ShuffleMode::T32_2X16_LO: // 16
+  case ShuffleMode::T32_2X16_HI: // 17
+  case ShuffleMode::T32_8X4_LO:  // 30
+  case ShuffleMode::T32_8X4_HI:  // 31
+  case ShuffleMode::T32_4X8_LO:  // 32
+  case ShuffleMode::T32_4X8_HI:  // 33
+    modeBitWidth = 32u;
+    break;
+  case ShuffleMode::T64_8X2_LO: //  6
+  case ShuffleMode::T64_8X2_HI: //  7
+  case ShuffleMode::T64_2X8_LO: // 14
+  case ShuffleMode::T64_2X8_HI: // 15
+    modeBitWidth = 64u;
+    break;
+  case ShuffleMode::T128_4X2_LO: //  8
+  case ShuffleMode::T128_4X2_HI: //  9
+  case ShuffleMode::T128_2X4_LO: // 12
+  case ShuffleMode::T128_2X4_HI: // 13
+    modeBitWidth = 128u;
+    break;
+  case ShuffleMode::T256_2X2_LO: // 10
+  case ShuffleMode::T256_2X2_HI: // 11
+    modeBitWidth = 256u;
+    break;
+  case ShuffleMode::T512_1X2_LO: // 22
+  case ShuffleMode::T512_1X2_HI: // 23
+    modeBitWidth = 512u;
+    break;
+  }
+
+  // Verify number of operands
+  if (requireRhs && !getRhs())
+    return emitError() << "shuffle mode '" << stringifyEnum(mode)
+                       << "' requires a second operand";
+
+  if (!requireRhs && getRhs())
+    return emitError() << "shuffle mode '" << stringifyEnum(mode)
+                       << "' does not admit a second operand";
+
+  // Verify vector element type
+  auto elemBitWidth =
+      cast<VectorType>(getLhs().getType()).getElementTypeBitWidth();
+  if (modeBitWidth != elemBitWidth)
+    return emitError() << "shuffle mode '" << stringifyEnum(mode)
+                       << "' requires vectors of " << modeBitWidth
+                       << "-bit elements";
+
```

Comment:
Should we also check if the operands and results are in 512-bit vectors?

Reply:
That's being checked by the type constraints in the op definition, this exclusively verifies whether the shuffle mode is "compatible" with the element type. It's up for debate whether we want to impose this restriction or not, since the hardware is agnostic to the mode (it just shuffles bytes around), but I'd rather be conservative for the time being.